File: sdk/memory-core/typescript/src/v3/wiki-client.ts
/**
 * v3 WikiClient — thin wrapper around Knowledge Service POST /v3/wiki/* (15 endpoints).
 *
 * Asset layer: create / get / list / delete / update-meta / ingest.
 * File layer: raw/{ls,read,write,rm} + page/{ls,read,write,rm}.
 * Derived: graph / search.
 *
 * Auth: x-tdai-service-id always (V3HttpTransport); Bearer only if apiKey
 * given (KS standalone needs none). Team-scoped writes carry team_id
 * (+ optional user_id) in body; id-only reads address by wiki_id.
 *
 * End-to-end wiki flow with MetadataClient + SkillClient:
 *   meta.createUser / createTeam / addTeamMember (management plane :8420)
 *   skills.create (skill plane :8420)
 *   wiki.create (KS data plane, this client) -> rawWrite sources -> ingest
 *   meta.shareAssetWithTeam() to share wiki/skill assets within their team
 *   (asset_id === wiki_id / skill_id); meta.grantAcl for restricted sharing,
 *   meta.setAgentFixedAssets to allocate an asset to an agent.
 */
import { ParamError } from "../errors.js";
import { V3HttpTransport } from "./http.js";
import type { Transport } from "../client.js";
import { flatGraphOptions, requireNonEmpty, stripUndefined } from "./request-helpers.js";
import type {
  WikiBatchDeleteResult, WikiClientConfig, WikiDetail, WikiGraphData,
  WikiIngestResult, WikiListResult, WikiPageRmResult, WikiPageWriteResultItem,
  WikiPageReadItem, WikiRawReadItem, WikiRawRmResult, WikiRawWriteItem,
  WikiSearchResult, PageEntry, RawFileEntry,
} from "./wiki-types.js";
const V3 = "/v3/wiki";
function need(name: string, v: string | undefined): string {
  try {
    return requireNonEmpty(name, v);
  } catch {
    throw new ParamError("WikiClient requires non-empty " + name);
  }
}
export class WikiClient {
  private readonly http: Transport;
  constructor(config: WikiClientConfig);
  constructor(transport: Transport);
  constructor(configOrTransport: WikiClientConfig | Transport) {
    if ("post" in configOrTransport) { this.http = configOrTransport; return; }
    const cfg = configOrTransport;
    need("serviceId", cfg.serviceId);
    if (!cfg.endpoint || !cfg.endpoint.trim()) throw new ParamError("WikiClient requires non-empty endpoint");
    this.http = new V3HttpTransport({
      endpoint: cfg.endpoint, apiKey: cfg.apiKey, serviceId: cfg.serviceId,
      userKey: cfg.userKey, timeout: cfg.timeout, rejectUnauthorized: cfg.rejectUnauthorized,
    });
  }
  create(p: { team_id: string; name: string; user_id?: string }): Promise<WikiDetail> {
    need("team_id", p.team_id); need("name", p.name);
    return this.http.post(V3 + "/create", stripUndefined({ team_id: p.team_id, name: p.name, user_id: p.user_id }));
  }
  get(wikiId: string): Promise<WikiDetail> {
    return this.http.post(V3 + "/get", { wiki_id: need("wiki_id", wikiId) });
  }
  list(teamId: string, opts: { status?: string; limit?: number; offset?: number } = {}): Promise<WikiListResult> {
    return this.http.post(V3 + "/list", stripUndefined({ team_id: need("team_id", teamId), ...opts }));
  }
  delete(wikiIds: string[]): Promise<WikiBatchDeleteResult> {
    if (!wikiIds.length) throw new ParamError("delete requires non-empty wiki_ids");
    return this.http.post(V3 + "/delete", { wiki_ids: wikiIds });
  }
  updateMeta(wikiId: string, patch: { name?: string; summary?: string | null }): Promise<WikiDetail> {
    if (!patch.name && patch.summary === undefined) throw new ParamError("updateMeta requires name or summary");
    return this.http.post(V3 + "/update-meta", stripUndefined({ wiki_id: need("wiki_id", wikiId), ...patch }));
  }
  rawWrite(teamId: string, wikiId: string, files: Array<{ filename: string; content: string }>, userId?: string): Promise<{ items: WikiRawWriteItem[] }> {
    need("team_id", teamId); need("wiki_id", wikiId);
    if (!files.length) throw new ParamError("rawWrite requires non-empty files");
    return this.http.post(V3 + "/raw/write", stripUndefined({ team_id: teamId, user_id: userId, wiki_id: wikiId, files }));
  }
  rawLs(wikiId: string): Promise<{ items: RawFileEntry[] }> {
    return this.http.post(V3 + "/raw/ls", { wiki_id: need("wiki_id", wikiId) });
  }
  rawRead(wikiId: string, filenames: string[]): Promise<{ items: WikiRawReadItem[] }> {
    if (!filenames.length) throw new ParamError("rawRead requires non-empty filenames");
    return this.http.post(V3 + "/raw/read", { wiki_id: need("wiki_id", wikiId), filenames });
  }
  rawRm(teamId: string, wikiId: string, filenames: string[], userId?: string): Promise<WikiRawRmResult> {
    if (!filenames.length) throw new ParamError("rawRm requires non-empty filenames");
    return this.http.post(V3 + "/raw/rm", stripUndefined({ team_id: need("team_id", teamId), user_id: userId, wiki_id: need("wiki_id", wikiId), filenames }));
  }
  pageLs(wikiId: string): Promise<{ items: PageEntry[] }> {
    return this.http.post(V3 + "/page/ls", { wiki_id: need("wiki_id", wikiId) });
  }
  pageRead(wikiId: string, refs: string[]): Promise<{ items: WikiPageReadItem[] }> {
    if (!refs.length) throw new ParamError("pageRead requires non-empty refs");
    return this.http.post(V3 + "/page/read", { wiki_id: need("wiki_id", wikiId), refs });
  }
  pageWrite(teamId: string, wikiId: string, pages: Array<{ ref: string; content: string }>, userId?: string): Promise<{ items: WikiPageWriteResultItem[] }> {
    if (!pages.length) throw new ParamError("pageWrite requires non-empty pages");
    return this.http.post(V3 + "/page/write", stripUndefined({ team_id: need("team_id", teamId), user_id: userId, wiki_id: need("wiki_id", wikiId), pages }));
  }
  pageRm(teamId: string, wikiId: string, refs: string[], userId?: string): Promise<WikiPageRmResult> {
    if (!refs.length) throw new ParamError("pageRm requires non-empty refs");
    return this.http.post(V3 + "/page/rm", stripUndefined({ team_id: need("team_id", teamId), user_id: userId, wiki_id: need("wiki_id", wikiId), refs }));
  }
  ingest(wikiId: string, userId?: string): Promise<WikiIngestResult> {
    return this.http.post(V3 + "/ingest", stripUndefined({ wiki_id: need("wiki_id", wikiId), user_id: userId }));
  }
  graph(wikiId: string): Promise<WikiGraphData> {
    return this.http.post(V3 + "/graph", { wiki_id: need("wiki_id", wikiId) });
  }
  /**
   * BM25 search with optional graph expansion. Options are sent as flat
   * top-level fields (hop/decay/minScore) — the shape the server reads.
   */
  search(wikiId: string, query: string, limit = 20, graph?: { hop?: number; decay?: number; minScore?: number }): Promise<WikiSearchResult> {
    need("query", query);
    return this.http.post(V3 + "/search", stripUndefined({
      wiki_id: need("wiki_id", wikiId), query, limit, ...flatGraphOptions(graph),
    }));
  }
}

File: sdk/memory-core/python/tencentdb_agent_memory/v3/wiki_client.py
"""TencentDB Agent Memory v3 Wiki SDK — Knowledge Service data-plane ``/v3/wiki/*`` (15 endpoints).

Asset layer: create / get / list / delete / update-meta / ingest.
File layer: raw/{ls,read,write,rm} + page/{ls,read,write,rm}.
Derived: graph / search (graph-expansion options go flat: hop/decay/minScore).

Auth: ``x-tdai-service-id`` always (shared ``HttpStub``); ``Bearer`` only
when ``api_key`` is given (KS standalone needs none). Team-scoped writes
carry ``team_id`` (+ optional ``user_id``); id-only reads address by ``wiki_id``.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from ..errors import ParamError
from ._common import flat_graph_options, ks_stub, need, strip_none

_V3 = "/v3/wiki"


class _WikiMethodsMixin:
    _stub: Any

    def _create(self, team_id: str, name: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/create", strip_none({"team_id": need("team_id", team_id), "name": need("name", name), "user_id": user_id}))

    def _get(self, wiki_id: str) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/get", {"wiki_id": need("wiki_id", wiki_id)})

    def _list(self, team_id: str, status: Optional[str] = None, limit: Optional[int] = None, offset: Optional[int] = None) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/list", strip_none({"team_id": need("team_id", team_id), "status": status, "limit": limit, "offset": offset}))

    def _delete(self, wiki_ids: List[str]) -> Dict[str, Any]:
        if not wiki_ids:
            raise ParamError("delete requires non-empty wiki_ids")
        return self._stub.post(f"{_V3}/delete", {"wiki_ids": wiki_ids})

    def _update_meta(self, wiki_id: str, name: Optional[str] = None, summary: Optional[str] = None) -> Dict[str, Any]:
        if not name and summary is None:
            raise ParamError("update_meta requires name or summary")
        return self._stub.post(f"{_V3}/update-meta", strip_none({"wiki_id": need("wiki_id", wiki_id), "name": name, "summary": summary}))

    def _raw_write(self, team_id: str, wiki_id: str, files: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        if not files:
            raise ParamError("raw_write requires non-empty files")
        return self._stub.post(f"{_V3}/raw/write", strip_none({"team_id": need("team_id", team_id), "user_id": user_id, "wiki_id": need("wiki_id", wiki_id), "files": files}))

    def _raw_ls(self, wiki_id: str) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/raw/ls", {"wiki_id": need("wiki_id", wiki_id)})

    def _raw_read(self, wiki_id: str, filenames: List[str]) -> Dict[str, Any]:
        if not filenames:
            raise ParamError("raw_read requires non-empty filenames")
        return self._stub.post(f"{_V3}/raw/read", {"wiki_id": need("wiki_id", wiki_id), "filenames": filenames})

    def _raw_rm(self, team_id: str, wiki_id: str, filenames: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        if not filenames:
            raise ParamError("raw_rm requires non-empty filenames")
        return self._stub.post(f"{_V3}/raw/rm", strip_none({"team_id": need("team_id", team_id), "user_id": user_id, "wiki_id": need("wiki_id", wiki_id), "filenames": filenames}))

    def _page_ls(self, wiki_id: str) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/page/ls", {"wiki_id": need("wiki_id", wiki_id)})

    def _page_read(self, wiki_id: str, refs: List[str]) -> Dict[str, Any]:
        if not refs:
            raise ParamError("page_read requires non-empty refs")
        return self._stub.post(f"{_V3}/page/read", {"wiki_id": need("wiki_id", wiki_id), "refs": refs})

    def _page_write(self, team_id: str, wiki_id: str, pages: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        if not pages:
            raise ParamError("page_write requires non-empty pages")
        return self._stub.post(f"{_V3}/page/write", strip_none({"team_id": need("team_id", team_id), "user_id": user_id, "wiki_id": need("wiki_id", wiki_id), "pages": pages}))

    def _page_rm(self, team_id: str, wiki_id: str, refs: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        if not refs:
            raise ParamError("page_rm requires non-empty refs")
        return self._stub.post(f"{_V3}/page/rm", strip_none({"team_id": need("team_id", team_id), "user_id": user_id, "wiki_id": need("wiki_id", wiki_id), "refs": refs}))

    def _ingest(self, wiki_id: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/ingest", strip_none({"wiki_id": need("wiki_id", wiki_id), "user_id": user_id}))

    def _graph(self, wiki_id: str) -> Dict[str, Any]:
        return self._stub.post(f"{_V3}/graph", {"wiki_id": need("wiki_id", wiki_id)})

    def _search(self, wiki_id: str, query: str, limit: int = 20, graph: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        need("query", query)
        body: Dict[str, Any] = {"wiki_id": need("wiki_id", wiki_id), "query": query, "limit": limit}
        body.update(flat_graph_options(graph))
        return self._stub.post(f"{_V3}/search", body)


class WikiClient(_WikiMethodsMixin):
    def __init__(self, endpoint: str = "", service_id: Optional[str] = None, *, api_key: str = "", user_key: Optional[str] = None, timeout: float = 30, verify: bool = True, stub: Optional[Any] = None) -> None:
        self._stub = stub if stub is not None else ks_stub(endpoint, service_id, api_key=api_key, timeout=timeout, verify=verify, user_key=user_key)

    def create(self, team_id: str, name: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._create(team_id, name, user_id)

    def get(self, wiki_id: str) -> Dict[str, Any]:
        return self._get(wiki_id)

    def list(self, team_id: str, status: Optional[str] = None, limit: Optional[int] = None, offset: Optional[int] = None) -> Dict[str, Any]:
        return self._list(team_id, status, limit, offset)

    def delete(self, wiki_ids: List[str]) -> Dict[str, Any]:
        return self._delete(wiki_ids)

    def update_meta(self, wiki_id: str, name: Optional[str] = None, summary: Optional[str] = None) -> Dict[str, Any]:
        return self._update_meta(wiki_id, name, summary)

    def raw_write(self, team_id: str, wiki_id: str, files: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._raw_write(team_id, wiki_id, files, user_id)

    def raw_ls(self, wiki_id: str) -> Dict[str, Any]:
        return self._raw_ls(wiki_id)

    def raw_read(self, wiki_id: str, filenames: List[str]) -> Dict[str, Any]:
        return self._raw_read(wiki_id, filenames)

    def raw_rm(self, team_id: str, wiki_id: str, filenames: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._raw_rm(team_id, wiki_id, filenames, user_id)

    def page_ls(self, wiki_id: str) -> Dict[str, Any]:
        return self._page_ls(wiki_id)

    def page_read(self, wiki_id: str, refs: List[str]) -> Dict[str, Any]:
        return self._page_read(wiki_id, refs)

    def page_write(self, team_id: str, wiki_id: str, pages: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._page_write(team_id, wiki_id, pages, user_id)

    def page_rm(self, team_id: str, wiki_id: str, refs: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._page_rm(team_id, wiki_id, refs, user_id)

    def ingest(self, wiki_id: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return self._ingest(wiki_id, user_id)

    def graph(self, wiki_id: str) -> Dict[str, Any]:
        return self._graph(wiki_id)

    def search(self, wiki_id: str, query: str, limit: int = 20, graph: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._search(wiki_id, query, limit, graph)

    def close(self) -> None:
        c = getattr(self._stub, "close", None)
        if callable(c):
            c()

    def __enter__(self) -> "WikiClient":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()


class AsyncWikiClient(_WikiMethodsMixin):
    def __init__(self, endpoint: str = "", service_id: Optional[str] = None, *, api_key: str = "", user_key: Optional[str] = None, timeout: float = 30, verify: bool = True, stub: Optional[Any] = None) -> None:
        self._stub = stub if stub is not None else ks_stub(endpoint, service_id, api_key=api_key, timeout=timeout, verify=verify, user_key=user_key, async_=True)

    async def create(self, team_id: str, name: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._create(team_id, name, user_id)

    async def get(self, wiki_id: str) -> Dict[str, Any]:
        return await self._get(wiki_id)

    async def list(self, team_id: str, status: Optional[str] = None, limit: Optional[int] = None, offset: Optional[int] = None) -> Dict[str, Any]:
        return await self._list(team_id, status, limit, offset)

    async def delete(self, wiki_ids: List[str]) -> Dict[str, Any]:
        return await self._delete(wiki_ids)

    async def update_meta(self, wiki_id: str, name: Optional[str] = None, summary: Optional[str] = None) -> Dict[str, Any]:
        return await self._update_meta(wiki_id, name, summary)

    async def raw_write(self, team_id: str, wiki_id: str, files: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._raw_write(team_id, wiki_id, files, user_id)

    async def raw_ls(self, wiki_id: str) -> Dict[str, Any]:
        return await self._raw_ls(wiki_id)

    async def raw_read(self, wiki_id: str, filenames: List[str]) -> Dict[str, Any]:
        return await self._raw_read(wiki_id, filenames)

    async def raw_rm(self, team_id: str, wiki_id: str, filenames: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._raw_rm(team_id, wiki_id, filenames, user_id)

    async def page_ls(self, wiki_id: str) -> Dict[str, Any]:
        return await self._page_ls(wiki_id)

    async def page_read(self, wiki_id: str, refs: List[str]) -> Dict[str, Any]:
        return await self._page_read(wiki_id, refs)

    async def page_write(self, team_id: str, wiki_id: str, pages: List[Dict[str, Any]], user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._page_write(team_id, wiki_id, pages, user_id)

    async def page_rm(self, team_id: str, wiki_id: str, refs: List[str], user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._page_rm(team_id, wiki_id, refs, user_id)

    async def ingest(self, wiki_id: str, user_id: Optional[str] = None) -> Dict[str, Any]:
        return await self._ingest(wiki_id, user_id)

    async def graph(self, wiki_id: str) -> Dict[str, Any]:
        return await self._graph(wiki_id)

    async def search(self, wiki_id: str, query: str, limit: int = 20, graph: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return await self._search(wiki_id, query, limit, graph)
